Makefile: add -Wextra flag and fix errors

To make compiler more strict we can add `-Wextra` and fix these errors:
```
~/d/beanstalkd (master *%=) % make -k
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o conn.o conn.c
conn.c:17:13: error: unused parameter 'a' [-Werror,-Wunused-parameter]
on_watch(ms a, tube t, size_t i)
            ^
conn.c:17:31: error: unused parameter 'i' [-Werror,-Wunused-parameter]
on_watch(ms a, tube t, size_t i)
                              ^
conn.c:24:14: error: unused parameter 'a' [-Werror,-Wunused-parameter]
on_ignore(ms a, tube t, size_t i)
             ^
conn.c:24:32: error: unused parameter 'i' [-Werror,-Wunused-parameter]
on_ignore(ms a, tube t, size_t i)
                               ^
4 errors generated.
make: *** [conn.o] Error 1
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o file.o file.c
file.c:204:30: error: comparison of integers of different signs: 'int32' (aka 'int') and 'size_t' (aka 'unsigned long') [-Werror,-Wsign-compare]
            if (jr.body_size > job_data_size_limit) {
                ~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~
file.c:328:30: error: comparison of integers of different signs: 'int32' (aka 'int') and 'size_t' (aka 'unsigned long') [-Werror,-Wsign-compare]
            if (jr.body_size > job_data_size_limit) {
                ~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~
file.c:463:11: error: comparison of integers of different signs: 'int' and 'unsigned long' [-Werror,-Wsign-compare]
    if (n < sizeof(int)) {
        ~ ^ ~~~~~~~~~~~
3 errors generated.
make: *** [file.o] Error 1
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o heap.o heap.c
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o job.o job.c
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o ms.o ms.c
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o net.o net.c
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o primes.o primes.c
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o prot.o prot.c
prot.c:873:39: error: missing field 'ru_maxrss' initializer [-Werror,-Wmissing-field-initializers]
    struct rusage ru = {{0, 0}, {0, 0}};
                                      ^
prot.c:1199:24: error: comparison of integers of different signs: 'unsigned long' and 'int' [-Werror,-Wsign-compare]
    if (strlen(c->cmd) != c->cmd_len - 2) {
        ~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~
prot.c:1557:23: error: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Werror,-Wsign-compare]
        for (i = 0; i < c->watch.used; i++) {
                    ~ ^ ~~~~~~~~~~~~~
prot.c:1650:22: error: unused parameter 'sig' [-Werror,-Wunused-parameter]
enter_drain_mode(int sig)
                     ^
prot.c:1878:19: error: comparison of integers of different signs: 'int' and 'size_t' (aka 'unsigned long') [-Werror,-Wsign-compare]
    for (i = 0; i < tubes.used; i++) {
                ~ ^ ~~~~~~~~~~
prot.c:1908:36: error: unused parameter 'which' [-Werror,-Wunused-parameter]
h_accept(const int fd, const short which, Server *s)
                                   ^
6 errors generated.
make: *** [prot.o] Error 1
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o sd-daemon.o sd-daemon.c
sd-daemon.c:46:23: error: unused parameter 'unset_environment' [-Werror,-Wunused-parameter]
int sd_listen_fds(int unset_environment) {
                      ^
sd-daemon.c:327:19: error: unused parameter 'unset_environment' [-Werror,-Wunused-parameter]
int sd_notify(int unset_environment, const char *state) {
                  ^
sd-daemon.c:327:50: error: unused parameter 'state' [-Werror,-Wunused-parameter]
int sd_notify(int unset_environment, const char *state) {
                                                 ^
sd-daemon.c:395:20: error: unused parameter 'unset_environment' [-Werror,-Wunused-parameter]
int sd_notifyf(int unset_environment, const char *format, ...) {
                   ^
sd-daemon.c:395:51: error: unused parameter 'format' [-Werror,-Wunused-parameter]
int sd_notifyf(int unset_environment, const char *format, ...) {
                                                  ^
5 errors generated.
make: *** [sd-daemon.o] Error 1
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o serv.o serv.c
serv.c:12:5: error: missing field 'use' initializer [-Werror,-Wmissing-field-initializers]
    },
    ^
serv.c:13:1: error: missing field 'sock' initializer [-Werror,-Wmissing-field-initializers]
};
^
2 errors generated.
make: *** [serv.o] Error 1
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o time.o time.c
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o tube.o tube.c
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o util.o util.c
./verc.sh >vers.c
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o vers.o vers.c
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o walg.o walg.c
walg.c:391:27: error: unused parameter 'j' [-Werror,-Wunused-parameter]
walresvupdate(Wal *w, job j)
                          ^
1 error generated.
make: *** [walg.o] Error 1
cc -Wall -Werror -Wextra -Wformat=2 -g   -c -o main.o main.c
main.c:72:10: error: unused parameter 'argc' [-Werror,-Wunused-parameter]
main(int argc, char **argv)
         ^
1 error generated.
make: *** [main.o] Error 1
make: Target `all' not remade because of errors.
```

Clang 10 on my OSX is not very strict for -Wextra. 
Errors produced in travis matrix:

**linux-clang:**
```
clang -Wall -Werror -Wextra -Wformat=2 -g   -c -o ct/_ctcheck.o ct/_ctcheck.c
ct/_ctcheck.c:135:7: error: missing field 'name' initializer [-Werror,-Wmissing-field-initializers]
    {0},
      ^
ct/_ctcheck.c:145:7: error: missing field 'name' initializer [-Werror,-Wmissing-field-initializers]
    {0},
      ^
2 errors generated.
```
**linux-gcc:**
```
gcc -O0 -ggdb -fprofile-arcs -ftest-coverage -Wall -Werror -Wextra -Wformat=2 -g   -c -o testserv.o testserv.c
testserv.c: In function ‘mustdiallocal’:
testserv.c:63:12: error: missing initializer for field ‘sin_family’ of ‘struct sockaddr_in’ [-Werror=missing-field-initializers]
     struct sockaddr_in addr = {};
            ^
In file included from /usr/include/x86_64-linux-gnu/bits/socket.h:146:0,
                 from /usr/include/x86_64-linux-gnu/sys/socket.h:38,
                 from testserv.c:9:
/usr/include/netinet/in.h:239:5: note: ‘sin_family’ declared here
     __SOCKADDR_COMMON (sin_);
     ^
testserv.c: In function ‘mustforksrv’:
testserv.c:170:16: error: missing initializer for field ‘r’ of ‘struct job’ [-Werror=missing-field-initializers]
         struct job list = {};
                ^
In file included from testserv.c:19:0:
dat.h:143:12: note: ‘r’ declared here
     Jobrec r; // persistent fields; these get written to the wal
            ^
cc1: all warnings being treated as errors
```

--- a/dat.h
+++ b/dat.h
@@ -31,7 +31,7 @@ typedef struct Wal    Wal;
 typedef void(*ms_event_fn)(ms a, void *item, size_t i);
 typedef void(*Handle)(void*, int rw);
 typedef int(*Less)(void*, void*);
-typedef void(*Record)(void*, int);
+typedef void(*Record)(void*, size_t);
 typedef int(FAlloc)(int, int);
 
 #if _LP64
@@ -81,14 +81,14 @@ struct stats {
 
 
 struct Heap {
-    int     cap;
-    int     len;
+    size_t  cap;
+    size_t  len;
     void    **data;
     Less    less;
     Record  rec;
 };
 int   heapinsert(Heap *h, void *x);
-void* heapremove(Heap *h, int k);
+void* heapremove(Heap *h, size_t k);
 
 
 struct Socket {
@@ -210,7 +210,7 @@ void job_free(job j);
 job job_find(uint64 job_id);
 
 /* the void* parameters are really job pointers */
-void job_setheappos(void*, int);
+void job_setheappos(void*, size_t);
 int job_pri_less(void*, void*);
 int job_delay_less(void*, void*);
 

--- a/heap.c
+++ b/heap.c
@@ -5,7 +5,7 @@
 
 
 static void
-set(Heap *h, int k, void *x)
+set(Heap *h, size_t k, void *x)
 {
     h->data[k] = x;
     h->rec(x, k);
@@ -13,7 +13,7 @@ set(Heap *h, int k, void *x)
 
 
 static void
-swap(Heap *h, int a, int b)
+swap(Heap *h, size_t a, size_t b)
 {
     void *tmp;
 
@@ -24,17 +24,17 @@ swap(Heap *h, int a, int b)
 
 
 static int
-less(Heap *h, int a, int b)
+less(Heap *h, size_t a, size_t b)
 {
     return h->less(h->data[a], h->data[b]);
 }
 
 
 static void
-siftdown(Heap *h, int k)
+siftdown(Heap *h, size_t k)
 {
     for (;;) {
-        int p = (k-1) / 2; /* parent */
+        size_t p = (k-1) / 2; /* parent */
 
         if (k == 0 || less(h, p, k)) {
             return;
@@ -47,16 +47,14 @@ siftdown(Heap *h, int k)
 
 
 static void
-siftup(Heap *h, int k)
+siftup(Heap *h, size_t k)
 {
     for (;;) {
-        int l, r, s;
-
-        l = k*2 + 1; /* left child */
-        r = k*2 + 2; /* right child */
+        size_t l = k*2 + 1; /* left child */
+        size_t r = k*2 + 2; /* right child */
 
         /* find the smallest of the three */
-        s = k;
+        size_t s = k;
         if (l < h->len && less(h, l, s)) s = l;
         if (r < h->len && less(h, r, s)) s = r;
 
@@ -75,24 +73,22 @@ siftup(Heap *h, int k)
 int
 heapinsert(Heap *h, void *x)
 {
-    int k;
-
     if (h->len == h->cap) {
         void **ndata;
-        int ncap = (h->len+1) * 2; /* allocate twice what we need */
+        size_t ncap = (h->len+1) * 2; /* allocate twice what we need */
 
         ndata = malloc(sizeof(void*) * ncap);
         if (!ndata) {
             return 0;
         }
 
-        memcpy(ndata, h->data, sizeof(void*)*h->len);
+        memcpy(ndata, h->data, sizeof(void*) * h->len);
         free(h->data);
         h->data = ndata;
         h->cap = ncap;
     }
 
-    k = h->len;
+    size_t k = h->len;
     h->len++;
     set(h, k, x);
     siftdown(h, k);
@@ -101,19 +97,16 @@ heapinsert(Heap *h, void *x)
 
 
 void *
-heapremove(Heap *h, int k)
+heapremove(Heap *h, size_t k)
 {
-    void *x;
-
     if (k >= h->len) {
         return 0;
     }
 
-    x = h->data[k];
+    void *x = h->data[k];
     h->len--;
     set(h, k, h->data[h->len]);
     siftdown(h, k);
     siftup(h, k);
-    h->rec(x, -1);
     return x;
 }

--- a/job.c
+++ b/job.c
@@ -158,7 +158,7 @@ job_free(job j)
 }
 
 void
-job_setheappos(void *j, int pos)
+job_setheappos(void *j, size_t pos)
 {
     ((job)j)->heap_index = pos;
 }

--- a/prot.c
+++ b/prot.c
@@ -193,9 +193,9 @@ size_t job_data_size_limit = JOB_DATA_SIZE_LIMIT_DEFAULT;
 #define STATS_TUBE_FMT "---\n" \
     "name: %s\n" \
     "current-jobs-urgent: %u\n" \
-    "current-jobs-ready: %u\n" \
+    "current-jobs-ready: %zu\n" \
     "current-jobs-reserved: %u\n" \
-    "current-jobs-delayed: %u\n" \
+    "current-jobs-delayed: %zu\n" \
     "current-jobs-buried: %u\n" \
     "total-jobs: %" PRIu64 "\n" \
     "current-using: %u\n" \

--- a/testheap.c
+++ b/testheap.c
@@ -10,13 +10,12 @@
 void
 cttest_heap_insert_one()
 {
-    Heap h = {0};
-    job j;
+    Heap h = {
+        .less = job_pri_less,
+        .rec = job_setheappos,
+    };
 
-    h.less = job_pri_less;
-    h.rec = job_setheappos;
-
-    j = make_job(1, 0, 1, 0, 0);
+    job j = make_job(1, 0, 1, 0, 0);
     assertf(j, "allocate job");
 
     heapinsert(&h, j);
@@ -27,34 +26,31 @@ cttest_heap_insert_one()
 void
 cttest_heap_insert_and_remove_one()
 {
-    Heap h = {0};
-    int r;
-    job j, j1;
+    Heap h = {
+        .less = job_pri_less,
+        .rec = job_setheappos,
+    };
 
-    h.less = job_pri_less;
-    h.rec = job_setheappos;
-    j1 = make_job(1, 0, 1, 0, 0);
+    job j1 = make_job(1, 0, 1, 0, 0);
     assertf(j1, "allocate job");
 
-    r = heapinsert(&h, j1);
+    int r = heapinsert(&h, j1);
     assertf(r, "insert should succeed");
 
-    j = heapremove(&h, 0);
-    assertf(j == j1, "j1 should come back out");
+    job got = heapremove(&h, 0);
+    assertf(got == j1, "j1 should come back out");
     assertf(h.len == 0, "h should be empty.");
-    printf("j->heap_index is %zu\n", j->heap_index);
-    assertf(j->heap_index == -1, "j's heap index should be invalid");
 }
 
 void
 cttest_heap_priority()
 {
-    Heap h = {0};
-    int r;
+    Heap h = {
+        .less = job_pri_less,
+        .rec = job_setheappos,
+    };
     job j, j1, j2, j3;
 
-    h.less = job_pri_less;
-    h.rec = job_setheappos;
     j1 = make_job(1, 0, 1, 0, 0);
     j2 = make_job(2, 0, 1, 0, 0);
     j3 = make_job(3, 0, 1, 0, 0);
@@ -62,7 +58,7 @@ cttest_heap_priority()
     assertf(j2, "allocate job");
     assertf(j3, "allocate job");
 
-    r = heapinsert(&h, j2);
+    int r = heapinsert(&h, j2);
     assertf(r, "insert should succeed");
     assertf(j2->heap_index == 0, "should match");
 
@@ -93,12 +89,12 @@ cttest_heap_priority()
 void
 cttest_heap_fifo_property()
 {
-    Heap h = {0};
-    int r;
+    Heap h = {
+        .less = job_pri_less,
+        .rec = job_setheappos,
+    };
     job j, j3a, j3b, j3c;
 
-    h.less = job_pri_less;
-    h.rec = job_setheappos;
     j3a = make_job(3, 0, 1, 0, 0);
     j3b = make_job(3, 0, 1, 0, 0);
     j3c = make_job(3, 0, 1, 0, 0);
@@ -106,7 +102,7 @@ cttest_heap_fifo_property()
     assertf(j3b, "allocate job");
     assertf(j3c, "allocate job");
 
-    r = heapinsert(&h, j3a);
+    int r = heapinsert(&h, j3a);
     assertf(r, "insert should succeed");
     assertf(h.data[0] == j3a, "j3a should be in pos 0");
     assertf(j3a->heap_index == 0, "should match");
@@ -140,22 +136,22 @@ cttest_heap_fifo_property()
 void
 cttest_heap_many_jobs()
 {
-    Heap h = {0};
-    uint last_pri;
-    int r, i, n = 20;
+    Heap h = {
+        .less = job_pri_less,
+        .rec = job_setheappos,
+    };
+    const int n = 20;
     job j;
 
-    h.less = job_pri_less;
-    h.rec = job_setheappos;
-
+    int i;
     for (i = 0; i < n; i++) {
         j = make_job(1 + rand() % 8192, 0, 1, 0, 0);
         assertf(j, "allocation");
-        r = heapinsert(&h, j);
+        int r = heapinsert(&h, j);
         assertf(r, "heapinsert");
     }
 
-    last_pri = 0;
+    uint last_pri = 0;
     for (i = 0; i < n; i++) {
         j = heapremove(&h, 0);
         assertf(j->r.pri >= last_pri, "should come out in order");
@@ -166,19 +162,18 @@ cttest_heap_many_jobs()
 void
 cttest_heap_remove_k()
 {
-    Heap h = {0};
-    uint last_pri;
-    int r, i, c, n = 20;
-    job j;
+    Heap h = {
+        .less = job_pri_less,
+        .rec = job_setheappos,
+    };
+    const int n = 20;
 
-    h.less = job_pri_less;
-    h.rec = job_setheappos;
-
+    int c, i;
     for (c = 0; c < 50; c++) {
         for (i = 0; i < n; i++) {
-            j = make_job(1 + rand() % 8192, 0, 1, 0, 0);
+            job j = make_job(1 + rand() % 8192, 0, 1, 0, 0);
             assertf(j, "allocation");
-            r = heapinsert(&h, j);
+            int r = heapinsert(&h, j);
             assertf(r, "heapinsert");
         }
 
@@ -186,9 +181,9 @@ cttest_heap_remove_k()
         heapremove(&h, 25);
 
         /* now make sure the rest are still a valid heap */
-        last_pri = 0;
+        uint last_pri = 0;
         for (i = 1; i < n; i++) {
-            j = heapremove(&h, 0);
+            job j = heapremove(&h, 0);
             assertf(j->r.pri >= last_pri, "should come out in order");
             last_pri = j->r.pri;
         }
@@ -198,18 +193,18 @@ cttest_heap_remove_k()
 void
 ctbench_heap_insert(int n)
 {
-    job *j;
-    int i;
-    j = calloc(n, sizeof *j);
+    job *j = calloc(n, sizeof *j);
     assert(j);
+    int i;
     for (i = 0; i < n; i++) {
         j[i] = make_job(1, 0, 1, 0, 0);
         assert(j[i]);
         j[i]->r.pri = -j[i]->r.id;
     }
-    Heap h = {0};
-    h.less = job_pri_less;
-    h.rec = job_setheappos;
+    Heap h = {
+        .less = job_pri_less,
+        .rec = job_setheappos,
+    };
     ctresettimer();
     for (i = 0; i < n; i++) {
         heapinsert(&h, j[i]);
@@ -219,14 +214,14 @@ ctbench_heap_insert(int n)
 void
 ctbench_heap_remove(int n)
 {
-    Heap h = {0};
-    job j;
-    int i;
+    Heap h = {
+        .less = job_pri_less,
+        .rec = job_setheappos,
+    };
 
-    h.less = job_pri_less;
-    h.rec = job_setheappos;
+    int i;
     for (i = 0; i < n; i++) {
-        j = make_job(1, 0, 1, 0, 0);
+        job j = make_job(1, 0, 1, 0, 0);
         assertf(j, "allocate job");
         heapinsert(&h, j);
     }

--- a/testserv.c
+++ b/testserv.c
@@ -89,6 +89,7 @@ mustdiallocal(int port)
 static void
 exit_process(int signum)
 {
+    UNUSED_PARAMETER(signum);
     exit(0);
 }
 

